✨  Add validation to nested ObjectMeta fields

LuBingtan opened the pull request:
<!-- please add an icon to the title of this PR (see https://sigs.k8s.io/cluster-api/CONTRIBUTING.md#contributing-a-patch), and delete this line and similar ones -->
<!-- the icon will be either ⚠️ (:warning:, major or breaking changes), ✨ (:sparkles:, feature additions), 🐛 (:bug:, patch and bugfixes), 📖 (:book:, documentation or proposals), or 🌱 (:seedling:, minor or other) -->

**What this PR does / why we need it**:
Today we have no validation for nested ObjectMeta fields
**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #8231

fabriziopandini commented:
/ok-to-test

cc @ykakarap @killianmuldoon 
It would be great to have unit tests ensuring no regressions for this behavior in all the validation webhooks for the CR with nested [ObjectMeta fields](https://github.com/kubernetes-sigs/cluster-api/pull/8431#top)

LuBingtan commented:
> /ok-to-test
> 
> cc @ykakarap @killianmuldoon It would be great to have unit tests ensuring no regressions for this behavior in all the validation webhooks for the CR with nested [ObjectMeta fields](https://github.com/kubernetes-sigs/cluster-api/pull/8431#top)

Sure, I'll add some unit tests

killianmuldoon commented on api/v1beta1/common_validate.go:
```diff
@@ -0,0 +1,36 @@
+		metadata.Labels,
+		parent.Child("labels"),
+	)
+	allErrs = append(allErrs, metavalidation.ValidateAnnotations(
```
```suggestion
	allErrs = append(allErrs, metavalidation.ValidateAnnotations(
```

Let's import this as apivalidation

LuBingtan replied:
fixed

killianmuldoon commented on api/v1beta1/common_validate.go:
```diff
@@ -0,0 +1,36 @@
+)
+
+// Validate validates the labels and annotations in ObjectMeta.
+func (metadata *ObjectMeta) Validate(parent *field.Path) field.ErrorList {
```
Can we add `metadata` to the fieldpath here instead of needing to pass it in every time?

LuBingtan replied:
When `ObjectMeta` is used as a field within a structure, we could give it any name, even though we commonly name it as `metadata`.
Therefore, I think it's more reasonable to pass `metadata` as a parameter. What do you think?

killianmuldoon replied:
You're right - very good point

killianmuldoon reviewed (commented):
Looks good - I think it might be better to have this function as an internal implementation - but I don't feel too strongly about it. 

If it's moved internal it will have to be a standalone function and take clusterv1.ObjectMeta as a parameter.

LuBingtan commented on api/v1beta1/common_validate.go:
```diff
@@ -0,0 +1,36 @@
+/*
+Copyright 2021 The Kubernetes Authors.
```
fixed

LuBingtan commented:
/retest pull-cluster-api-test-main

LuBingtan commented:
/retest

LuBingtan commented on api/v1beta1/common_validate.go:
```diff
@@ -0,0 +1,36 @@
+/*
```
Thanks for your review!

I think it's better to have this function here rather than in the "internal" path, because we may need to call it in other places outside of this package. 
For example, here in the package `sigs.k8s.io/cluster-api/test` we also need to use this function for validation:
https://github.com/kubernetes-sigs/cluster-api/blob/f1129eb781ae2b002ae8afa7b0dc4a392ce3537f/test/infrastructure/docker/api/v1beta1/dockerclustertemplate_webhook.go#L66-L67

So, I think it's more appropriate to place this function here for now. What do you think?

killianmuldoon replied:
We should be able to call functions from the top-level internal package across the codebase, so I don't think it would interfere with the `test/infrastructure/docker` usage.

LuBingtan replied:
Oh, you are right. I didn't realize that earlier. I'll move this function to the internal package.

LuBingtan replied:
I found another problem.

The function itself requires the package `sigs.k8s.io/cluster-api/api/v1beta1` to be imported, and the webhook interface functions in this package also need to call the `Validate` function. If we put this function in a different package, it will result in a circular dependency between the two packages. 

Therefore, it seems that this function cannot be placed in the internal package.

sbueringer replied:
I would suggest to keep the function here for simplicity.

We also have a history of dropping Validate*() functions on API types. So if we have to get rid of it at some point there is precedent for deprecation + removal

LuBingtan commented:
Hi @killianmuldoon @fabriziopandini 
Thanks for your review, it's very helpful to me!
I have made some updates and have addressed most of the comments. 
And for some of the comments, I have provided different opinions and replied in the corresponding thread. Please take a look, thanks.

sbueringer commented on internal/webhooks/clusterclass.go:
```diff
@@ -156,6 +156,9 @@ func (webhook *ClusterClass) validate(ctx context.Context, oldClusterClass, newC
 	// Validate patches.
 	allErrs = append(allErrs, validatePatches(newClusterClass)...)
 
+	// Validate metadata
```
Let's also validate the two metadata fields below Cluster.spec.topology

sbueringer reviewed (commented):
Looks good only minor findings

LuBingtan commented:
> Looks good only minor findings
> 
> (sorry for the late review. I think when those two findings are adressed we can merge quickly)

@sbueringer  Thanks! Already updated, I believe all comments are addressed. Please take a look again.

sbueringer commented on internal/webhooks/clusterclass.go:
```diff
@@ -176,6 +179,10 @@ func (webhook *ClusterClass) validate(ctx context.Context, oldClusterClass, newC
 		// Ensure no MachineHealthCheck currently in use has been removed from the ClusterClass.
 		allErrs = append(allErrs,
 			validateUpdatesToMachineHealthCheckClasses(clusters, oldClusterClass, newClusterClass)...)
+
+		// Ensure Topology metadata in every clusters is valid.
+		allErrs = append(allErrs,
+			validateClustersTopologyMetadata(clusters)...)
```
```suggestion
```
I wouldn't add this here. I think it's enough if we validate clusters in the cluster webhook

LuBingtan replied:
Agree. removed this part

sbueringer commented:
Thank you very much!!

/lgtm
/approve

g-gaston commented:
/area api
